src/bin/pgcopydb/ld_transform.c

```diff
@@ -2609,3 +2650,323 @@ LogicalMessageValueEq(LogicalMessageValue *a, LogicalMessageValue *b)
+ */
+static bool
+findGeneratedColumns(GeneratedColumnsCache *cache,
+					 const char *nspname,
+					 const char *relname,
+					 GeneratedColumns **columns)
+{
+	/*
+	 * NULL cache means that we don't have any generated columns in the
+	 * catalog.
+	 */
+	if (cache == NULL)
+	{
+		return false;
+	}
+
+	GeneratedColumnsCache *item = NULL;
+
+	TableWithGeneratedColumns generatedColumn = { 0 };
+
+	NORMALIZE_IDENTIFIER(generatedColumn.nspname, nspname, PG_NAMEDATALEN);
+	NORMALIZE_IDENTIFIER(generatedColumn.relname, relname, PG_NAMEDATALEN);
+
+	HASH_FIND(hh, cache, &generatedColumn, sizeof(TableWithGeneratedColumns), item);
+
+	if (item == NULL)
+	{
+		return false;
+	}
+
+	if (item->columns == NULL)
+	{
+		log_error("BUG: Table \"%s.%s\" is in the cache but columns are NULL",
+				  nspname, relname);
+		return false;
+	}
+
+	*columns = item->columns;
+
+	log_debug("Table \"%s.%s\" has generated columns", nspname, relname);
+	return true;
+}
+
+
+/*
+ * isGeneratedColumn checks whether the given "attname" is a generated column.
+ *
+ * Returns true if the column is generated, false otherwise.
+ *
+ * NOTE: There is no error condition, if the cache is NULL, it means that we
+ * don't have any generated columns in the catalog.
+ */
+static bool
+isGeneratedColumn(GeneratedColumns *columns, const char *attname)
+{
+	char column[PG_NAMEDATALEN] = { 0 };
+
+	NORMALIZE_IDENTIFIER(column, attname, PG_NAMEDATALEN);
+
+	GeneratedColumns *generatedColumns = NULL;
+
+	HASH_FIND_STR(columns, column, generatedColumns);
+
+	if (generatedColumns != NULL)
+	{
+		log_debug("Column \"%s\" is generated", attname);
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * generated_column_iter_hook is a callback function that populates the
+ * generated columns cache from the catalog.
+ */
+static bool
+generated_column_iter_hook(void *ctx, SourceTable *table)
+{
+	StreamSpecs *specs = (StreamSpecs *) ctx;
+	StreamContext *privateContext = &(specs->private);
+	DatabaseCatalog *sourceDB = specs->sourceDB;
+
+	if (!catalog_s_table_fetch_attrs(sourceDB, table))
+	{
+		log_error("Failed to fetch attributes for table \"%s\".%s",
+				  table->nspname, table->relname);
+		return false;
+	}
+
+	GeneratedColumnsCache *item = (GeneratedColumnsCache *)
+								  calloc(1, sizeof(GeneratedColumnsCache));
+
+	if (item == NULL)
+	{
+		log_error(ALLOCATION_FAILED_ERROR);
+		return false;
+	}
+
+	for (int i = 0; i < table->attributes.count; i++)
+	{
+		SourceTableAttribute *attr = &(table->attributes.array[i]);
+
+		if (attr->attisgenerated)
+		{
+			/*
+			 * Add generated column to the cache.
+			 */
+			GeneratedColumns *generatedColumn = (GeneratedColumns *)
+												calloc(1, sizeof(GeneratedColumns));
+			if (generatedColumn == NULL)
+			{
+				log_error(ALLOCATION_FAILED_ERROR);
+				return false;
+			}
+
+			NORMALIZE_IDENTIFIER(generatedColumn->column,
+								 attr->attname, PG_NAMEDATALEN);
+			HASH_ADD_STR(item->columns, column, generatedColumn);
```

Comment:
We now have a hash-table per table with generated column. I believe the previous design had a single hash-table with all the generated columns in there, and I believe I prefer that option. How could you use the hash-table API that we have to maintain a single hash-table that would contain `nspname.relname` and also another single hash-table that would contain `nspname.relname.attname`?

Reply:
Or maybe the hash-table overhead is much less compared to having to repeat the nspname and relname in memory for each generated column?

Reply:
@dimitri, It depends on the nature of the schema. We can compromise both of the implementation using micro benchmarks. For e.g. 10 tables with each having 3 generated columns each vs 3 tables having 10 generated columns each.

I prefer the current approach, i.e. one hash table to find `nspname.relname` and another hash table to find `attname`.

Reply:
I can live with that too, but then it needs to be explained in the comments why that choice was made, and how we did conclude that the “one hash-table per table with generated columns” is an acceptable overhead.